ehrql/serializer.py

```diff
@@ -283,6 +334,19 @@ def unmarshal_for_object(self, type_, value):
         attrs = {key: self.unmarshal(v) for key, v in value.items()}
         return type_(**attrs)
 
+    @unmarshal_for.register(SelectTable)
+    @unmarshal_for.register(SelectPatientTable)
+    def unmarshal_for_prohibited(self, type_, value):
+        # We want any table nodes in the resulting query graph to be trustworthy (i.e.
+        # not user-supplied) so that any permission restrictions they impose can be
+        # relied upon. This means that such nodes can only be serialized as "external
+        # references" i.e. references to nodes defined outside of the user's query
+        # graph.
+        raise SerializerError(
```

Comment:
Just checking my understanding: Is this the main security barrier that prevents user code from bypassing permissions?

Would the query engine still understand/execute an inline table ref if the usr manage to return a serialized qm that contained one?

Reply:
Yeah, it is. If the user manages to get their own `SelectTable` or `SelectPatientTable` object into the graph then they could define it as having different permissions.

As a belt-and-braces we could add an extra assert/check in here:
https://github.com/opensafely-core/ehrql/blob/491cc911cb5d3f8170e34fa5e08e1b52bff5ffd8/ehrql/serializer.py#L282-L284

Could be as simple as:
```py
assert not issubclass(type_, SelectTable | SelectPatientTable)
```